pymc/tests/test_sampling.py

```diff
@@ -1446,3 +1449,171 @@ def test_no_init_nuts_compound(caplog):
         b = pm.Poisson("b", 1)
         pm.sample(10, tune=10)
         assert "Initializing NUTS" not in caplog.text
+
+
+class TestCompileForwardSampler:
+    @staticmethod
+    def get_function_roots(function):
+        roots = []
+
+        def expand(node):
+            if (
+                isinstance(node, Variable)
+                and node.owner is None
+                and getattr(node, "name", None) is not None
+            ):
+                roots.append(node)
+            return node.get_parents()
+
+        list(walk(function.maker.fgraph.outputs, expand))
+        return roots
+
+    @staticmethod
+    def get_function_inputs(function):
+        return {
+            i
+            for i, istore, *_ in function.indices
+            if istore is None and not isinstance(i.variable, SharedVariable)
+        }
```

Comment:
Yeah, that should work. I originally had tried to do something similar but used `fn` instead of `maker`, and that failed for some compile modes. I'll switch to your suggestion, which is cleaner.